sriov: add support for drivers-autoprobe

rjarry opened the pull request:
NetworkManager already supports setting sriov_drivers_autoprobe, add it in the nmstate schema.

I am not sure if what I am doing is enough here. I'd like some guidance.

Thanks!

rjarry committed:
sriov: add support for autoprobe-drivers

NetworkManager already supports setting sriov_drivers_autoprobe, add it
in the nmstate schema.

Add integration test to ensure that it works as expected.

Signed-off-by: Robin Jarry <rjarry@redhat.com>

cathay4t commented on examples/eth1_with_sriov.yml:
```diff
@@ -5,6 +5,7 @@ interfaces:
     state: up
     ethernet:
       sr-iov:
+        autoprobe-drivers: false
```
The kernel name for this is `sriov_drivers_autoprobe`, instead of using NetworkManager name, can we try `drivers_autoprobe`?

rjarry replied:
I have changed it to better match the kernel name.

cathay4t commented on rust/src/lib/ifaces/sriov.rs:
```diff
@@ -43,6 +44,14 @@ use crate::{
+        default,
+        deserialize_with = "crate::deserializer::option_bool_or_string"
+    )]
+    /// The number of VFs enabled on PF.
```
This document is wrong.

rjarry replied:
Fixed.

cathay4t reviewed (changes requested):
Please include integration test case.

You seems missing query code on this property, hence verification will fail.
Regarding this query, for now, you can just place it into `rust/src/lib/nispor/ethernet.rs`.
ideally, this query code should goes into [nispor crate][https://github.com/nispor/nispor/], not a must.

rjarry commented:
Hi @cathay4t, I have added an integration test and ran `make check` locally. I tried patching https://github.com/nispor/nispor but couldn't find any netlink reference for `sriov_drivers_autoprobe` so that may not be the best place to add it. Instead I added direct access to sysfs in nmstate. I hope that is OK.

I couldn't figure out how to run the integration tests by myself. Hopefully, the CI will do that.

Let me know if there is anything else.

cathay4t commented:
The sriov_drivers_autoprobe is for PCI domain where netlink is not their API. Please use sysfs instead.

cathay4t commented:
> Hi @cathay4t, I have added an integration test and ran `make check` locally. I tried patching https://github.com/nispor/nispor but couldn't find any netlink reference for `sriov_drivers_autoprobe` so that may not be the best place to add it. Instead I added direct access to sysfs in nmstate. I hope that is OK.
> 
> I couldn't figure out how to run the integration tests by myself. Hopefully, the CI will do that.
> 
> Let me know if there is anything else.

For integration test, you may mimic test cases in `sriov_test.py`.
To run the integration test, you may:
 1. `make rpm`.
 2. Installed the rpm just built.
 3. `cd tests/integration`
 4. `sudo env TEST_REAL_NIC=<your_sriov_nic> pytest -k <new_test_case>`

cathay4t commented:
The CI will not run sriov test as CI has no such hardwares, we have to run it manually on local hardwares.

cathay4t commented on rust/src/lib/nispor/ethernet.rs:
```diff
@@ -18,6 +19,26 @@ fn gen_eth_conf(np_iface: &nispor::Iface) -> EthernetConfig {
+                        if i == 0 { Some(false) } else { Some(true) };
+                }
+                Err(err) => {
+                    log::warn!("failed to parse {:?}: {:?}", &p, err);
```
Do not use debug out here, use `PathBuf::display()` and the error as Display trait implemented.

rjarry replied:
`PathBuf::display()` does not exist. Is that OK like so?

```rust
                    log::warn!(
                        "failed to parse {}: {:?}",
                        p.to_string_lossy(),
                        err
                    );
```

rjarry replied:
I have tested this and it works.

```
[2024-04-05T20:31:32Z WARN  nmstate::nispor::ethernet] failed to parse /sys/class/net/eno12409/device/sriov_drivers_autoprobe: ParseIntError { kind: InvalidDigit }
```

cathay4t reviewed (changes requested):
Only a small lint to fix.

Please also give me 3 days to find a sriov server to test this patch.

rjarry commented:
@cathay4t I have pushed a small fix and updated the logging message as requested.

The feature seems to work during manual testing. However, the sriov test fails. I must have missed something. I will have a fresh look next week.

<details>

```console
[root@dell-r450-nfv2023-01 ~/nmstate/tests/integration]# ethtool -i ens1f0np0
driver: mlx5_core
version: 5.14.0-424.el9.x86_64
firmware-version: 26.34.1002 (MT_0000000532)
expansion-rom-version: 
bus-info: 0000:18:00.0
supports-statistics: yes
supports-test: yes
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: yes
[root@dell-r450-nfv2023-01 ~/nmstate/tests/integration]# ip link show ens1f0np0
5: ens1f0np0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether b8:3f:d2:fa:53:86 brd ff:ff:ff:ff:ff:ff
    altname enp24s0f0np0
[root@dell-r450-nfv2023-01 ~/nmstate/tests/integration]# grep . /sys/class/net/ens1f0np0/device/sriov_*
/sys/class/net/ens1f0np0/device/sriov_drivers_autoprobe:1
/sys/class/net/ens1f0np0/device/sriov_numvfs:0
/sys/class/net/ens1f0np0/device/sriov_offset:2
/sys/class/net/ens1f0np0/device/sriov_stride:1
/sys/class/net/ens1f0np0/device/sriov_totalvfs:64
/sys/class/net/ens1f0np0/device/sriov_vf_device:101e
/sys/class/net/ens1f0np0/device/sriov_vf_total_msix:0
[root@dell-r450-nfv2023-01 ~/nmstate/tests/integration]# env TEST_REAL_NIC=ens1f0np0 pytest sriov_test.py
========================== test session starts =================================
platform linux -- Python 3.9.18, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
RPMs: NetworkManager-1.45.91-1.el9.x86_64
OS: Red Hat Enterprise Linux 9.4 Beta (Plow)
nmstate: nmstate-2.2.28-0.alpha.20240405.0ed825288482.el9.x86_64

rootdir: /root/nmstate, configfile: tox.ini
collected 20 items

sriov_test.py .........F...FF...FF                                        [100%]

============================== FAILURES ========================================
___________ TestSrIov.test_sriov_vf_vlan_id_and_qos_proto ______________________

self = <integration.sriov_test.TestSrIov object at 0x7f757cd574f0>

    def test_sriov_vf_vlan_id_and_qos_proto(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {
                            Ethernet.SRIOV.TOTAL_VFS: 2,
                            Ethernet.SRIOV.VFS_SUBTREE: [
                                {
                                    Ethernet.SRIOV.VFS.ID: 0,
                                    Ethernet.SRIOV.VFS.VLAN_ID: 100,
                                    Ethernet.SRIOV.VFS.QOS: 5,
                                    Ethernet.SRIOV.VFS.VLAN_PROTO: "802.1ad",
                                },
                                {
                                    Ethernet.SRIOV.VFS.ID: 1,
                                    Ethernet.SRIOV.VFS.VLAN_ID: 102,
                                    Ethernet.SRIOV.VFS.QOS: 6,
                                    Ethernet.SRIOV.VFS.VLAN_PROTO: "802.1q",
                                },
                            ],
                        }
                    },
                }
            ]
        }
        try:
>           libnmstate.apply(desired_state)

sriov_test.py:289: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/site-packages/libnmstate/netapplier.py:29: in apply
    return apply_net_state(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

state = {'interfaces': [{'ethernet': {'sr-iov': {'total-vfs': 2, 'vfs': [{...}, {...}]}}, 'name': 'ens1f0np0', 'state': 'absent'}]}, kernel_only = False, verify_change = True
save_to_disk = True, commit = True, rollback_timeout = 60

    def apply_net_state(
        state,
        kernel_only=False,
        verify_change=True,
        save_to_disk=True,
        commit=True,
        rollback_timeout=60,
    ):
        c_err_msg = c_char_p()
        c_err_kind = c_char_p()
        c_state = c_char_p(json.dumps(state).encode("utf-8"))
        c_log = c_char_p()
        flags = NMSTATE_FLAG_NONE
        if kernel_only:
            flags |= NMSTATE_FLAG_KERNEL_ONLY
    
        if not verify_change:
            flags |= NMSTATE_FLAG_NO_VERIFY
    
        if not commit:
            flags |= NMSTATE_FLAG_NO_COMMIT
    
        if not save_to_disk:
            flags |= NMSTATE_FLAG_MEMORY_ONLY
    
        rc = lib.nmstate_net_state_apply(
            flags,
            c_state,
            rollback_timeout,
            byref(c_log),
            byref(c_err_kind),
            byref(c_err_msg),
        )
        err_msg = c_err_msg.value
        err_kind = c_err_kind.value
        parse_log(c_log.value)
        lib.nmstate_cstring_free(c_log)
        lib.nmstate_cstring_free(c_err_kind)
        lib.nmstate_cstring_free(c_err_msg)
        if rc != NMSTATE_PASS:
>           raise map_error(err_kind, err_msg)
E           libnmstate.error.NmstateVerificationError: Verification failure: ens1f0np0.interface.ethernet.sr-iov.vfs[0].vlan-id desire '100', current '0'

/usr/lib/python3.9/site-packages/libnmstate/clib_wrapper.py:127: NmstateVerificationError
----------------------------- Captured log call --------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  root:clib_wrapper.py:273 1712349566:nmstate::nm::checkpoint:src/lib/nm/checkpoint.rs:27: Timeout:Timeout on waiting rollback
________TestSrIov.test_refer_vf_using_pf_name_and_vf_id_ovs_bridge _____________

self = <integration.sriov_test.TestSrIov object at 0x7f757c40a190>, sriov_interface = {'interfaces': [{'ethernet': {'sr-iov': {'total-vfs': 2}}, 'name': 'ens1f0np0', 'state': 'up'}]}

    def test_refer_vf_using_pf_name_and_vf_id_ovs_bridge(
        self, sriov_interface
    ):
        pf_name = _test_nic_name()
>       with ovs_bridge(
            TEST_BRIDGE,
            [f"sriov:{pf_name}:0", f"sriov:{pf_name}:1"],
            TEST_OVS_IFACE,
        ) as expected_state:

sriov_test.py:385: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib64/python3.9/contextlib.py:119: in __enter__
    return next(self.gen)
testlib/ovslib.py:159: in ovs_bridge
    with bridge.create():
/usr/lib64/python3.9/contextlib.py:119: in __enter__
    return next(self.gen)
testlib/ovslib.py:125: in create
    libnmstate.apply(desired_state)
/usr/lib/python3.9/site-packages/libnmstate/netapplier.py:29: in apply
    return apply_net_state(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

state = {'interfaces': [{'bridge': {'port': [{'name': 'test-ovs0'}, {'name': 'sriov:ens1f0np0:0'}, {'name': 'sriov:ens1f0np0:1...ame': 'test-br0', 'state': 'up', 'type': 'ovs-bridge'}, {'name': 'test-ovs0', 'state': 'up', 'type': 'ovs-interface'}]}
kernel_only = False, verify_change = True, save_to_disk = True, commit = True, rollback_timeout = 60

    def apply_net_state(
        state,
        kernel_only=False,
        verify_change=True,
        save_to_disk=True,
        commit=True,
        rollback_timeout=60,
    ):
        c_err_msg = c_char_p()
        c_err_kind = c_char_p()
        c_state = c_char_p(json.dumps(state).encode("utf-8"))
        c_log = c_char_p()
        flags = NMSTATE_FLAG_NONE
        if kernel_only:
            flags |= NMSTATE_FLAG_KERNEL_ONLY
    
        if not verify_change:
            flags |= NMSTATE_FLAG_NO_VERIFY
    
        if not commit:
            flags |= NMSTATE_FLAG_NO_COMMIT
    
        if not save_to_disk:
            flags |= NMSTATE_FLAG_MEMORY_ONLY
    
        rc = lib.nmstate_net_state_apply(
            flags,
            c_state,
            rollback_timeout,
            byref(c_log),
            byref(c_err_kind),
            byref(c_err_msg),
        )
        err_msg = c_err_msg.value
        err_kind = c_err_kind.value
        parse_log(c_log.value)
        lib.nmstate_cstring_free(c_log)
        lib.nmstate_cstring_free(c_err_kind)
        lib.nmstate_cstring_free(c_err_msg)
        if rc != NMSTATE_PASS:
>           raise map_error(err_kind, err_msg)
E           libnmstate.error.NmstateDependencyError: Manager(MissingPlugin):NetworkManager plugin for 'ovs-bridge' unavailable

/usr/lib/python3.9/site-packages/libnmstate/clib_wrapper.py:127: NmstateDependencyError
---------------------------Captured log setup ----------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
---------------------------Captured log teardown -------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
___________TestSrIov.test_refer_vf_using_pf_name_and_vf_id_ovs_bond ____________

self = <integration.sriov_test.TestSrIov object at 0x7f757cd4b700>, sriov_interface = {'interfaces': [{'ethernet': {'sr-iov': {'total-vfs': 2}}, 'name': 'ens1f0np0', 'state': 'up'}]}

    def test_refer_vf_using_pf_name_and_vf_id_ovs_bond(self, sriov_interface):
        pf_name = _test_nic_name()
>       with ovs_bridge_bond(
            TEST_BRIDGE,
            {TEST_BOND: [f"sriov:{pf_name}:0", f"sriov:{pf_name}:1"]},
            TEST_OVS_IFACE,
        ) as expected_state:

sriov_test.py:406: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib64/python3.9/contextlib.py:119: in __enter__
    return next(self.gen)
testlib/ovslib.py:171: in ovs_bridge_bond
    with bridge.create():
/usr/lib64/python3.9/contextlib.py:119: in __enter__
    return next(self.gen)
testlib/ovslib.py:125: in create
    libnmstate.apply(desired_state)
/usr/lib/python3.9/site-packages/libnmstate/netapplier.py:29: in apply
    return apply_net_state(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

state = {'interfaces': [{'bridge': {'port': [{'name': 'test-ovs0'}, {'link-aggregation': {...}, 'name': 'test-bond0'}]}, 'name': 'test-br0', 'state': 'up', 'type': 'ovs-bridge'}, {'name': 'test-ovs0', 'state': 'up', 'type': 'ovs-interface'}]}
kernel_only = False, verify_change = True, save_to_disk = True, commit = True, rollback_timeout = 60

    def apply_net_state(
        state,
        kernel_only=False,
        verify_change=True,
        save_to_disk=True,
        commit=True,
        rollback_timeout=60,
    ):
        c_err_msg = c_char_p()
        c_err_kind = c_char_p()
        c_state = c_char_p(json.dumps(state).encode("utf-8"))
        c_log = c_char_p()
        flags = NMSTATE_FLAG_NONE
        if kernel_only:
            flags |= NMSTATE_FLAG_KERNEL_ONLY
    
        if not verify_change:
            flags |= NMSTATE_FLAG_NO_VERIFY
    
        if not commit:
            flags |= NMSTATE_FLAG_NO_COMMIT
    
        if not save_to_disk:
            flags |= NMSTATE_FLAG_MEMORY_ONLY
    
        rc = lib.nmstate_net_state_apply(
            flags,
            c_state,
            rollback_timeout,
            byref(c_log),
            byref(c_err_kind),
            byref(c_err_msg),
        )
        err_msg = c_err_msg.value
        err_kind = c_err_kind.value
        parse_log(c_log.value)
        lib.nmstate_cstring_free(c_log)
        lib.nmstate_cstring_free(c_err_kind)
        lib.nmstate_cstring_free(c_err_msg)
        if rc != NMSTATE_PASS:
>           raise map_error(err_kind, err_msg)
E           libnmstate.error.NmstateDependencyError: Manager(MissingPlugin):NetworkManager plugin for 'ovs-bridge' unavailable

/usr/lib/python3.9/site-packages/libnmstate/clib_wrapper.py:127: NmstateDependencyError
----------------------------Captured log setup ---------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
-----------------------------Captured log call ---------------------------------
---------------------------Captured log teardown -------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
__________________TestSrIov.test_change_vf_parameters_only _____________________

self = <integration.sriov_test.TestSrIov object at 0x7f757cd577c0>, sriov_with_62_vfs = None

    def test_change_vf_parameters_only(self, sriov_with_62_vfs):
        pf_name = _test_nic_name()
        cur_iface = show_only((pf_name,))[Interface.KEY][0]
        iface_infos = [
            {
                Interface.NAME: pf_name,
                Interface.TYPE: InterfaceType.ETHERNET,
                Interface.STATE: InterfaceState.UP,
                Ethtool.CONFIG_SUBTREE: cur_iface[Ethtool.CONFIG_SUBTREE],
                Ethernet.CONFIG_SUBTREE: {
                    Ethernet.SRIOV_SUBTREE: {
                        Ethernet.SRIOV.VFS_SUBTREE: [VF0_CONF, VF1_CONF],
                    }
                },
            },
        ]
        desired_state = {Interface.KEY: iface_infos}
        libnmstate.apply(desired_state)
        vf_ifaces = get_sriov_vf_names(pf_name)
>       assert len(vf_ifaces) == 62
E       AssertionError: assert 2 == 62
E        +  where 2 = len(['ens1f0v0', 'ens1f0v1'])

sriov_test.py:528: AssertionError
----------------------------- Captured log setup -------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
------------------------------Captured log call --------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn1/net/: No such file or directory (os error 2)
---------------------------Captured log teardown -------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
____________________TestSrIov.test_drivers_autoprobe_false _____________________

self = <integration.sriov_test.TestSrIov object at 0x7f757cdf4d00>

    def test_drivers_autoprobe_false(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {
                            Ethernet.SRIOV.DRIVERS_AUTOPROBE: False,
                            Ethernet.SRIOV.TOTAL_VFS: 32,
                        },
                    },
                }
            ]
        }
        libnmstate.apply(desired_state)
        vf_ifaces = get_sriov_vf_names(pf_name)
>       assert len(vf_ifaces) == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = len(['ens1f0v1'])

sriov_test.py:547: AssertionError
-------------------------- Captured log call -----------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
------------------------ Captured log teardown ---------------------------------
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
WARNING  nispor-1.2.18/query/sriov.rs:186: Failed to read dir /sys/class/net/ens1f0np0/device/virtfn0/net/: No such file or directory (os error 2)
================= 5 failed, 15 passed in 275.55s (0:04:35) =====================
```
</details>

cathay4t commented:
Change required: Please skip `verify_sriov()` when `drivers-autoprobe: false` because there will be no NIC for VFs.

The `verify_sriov()` is a function to retry and wait on VF got its kernel NIC created, for this use case we do not need it.

I tried in real hardware with `nmstatectl apply --no-verify`, it works as expected. Let's see how the verification will be after you skipped the `verify_sriov()` when `drivers-autoprobe: false`

rjarry commented:
Ack I will try this.

Also I think that drivers-autoprobe should default to true even if unset to reflect the behavior of linux. If we don't do this we may have issues when setting drivers-autoprobe=false and then unsetting it.

liangwen12year commented on rust/src/lib/unit_tests/ethernet.rs:
```diff
@@ -15,6 +15,7 @@ ethernet:
   auto-negotiation: "false"
   speed: "1000"
   sr-iov:
+    drivers-autoprobe: "off"
```
Can use false instead off here ?

rjarry replied:
Will do.

liangwen12year commented on tests/integration/sriov_test.py:
```diff
@@ -95,7 +95,10 @@ def sriov_interface(disable_sriov):
-            Ethernet.SRIOV_SUBTREE: {Ethernet.SRIOV.TOTAL_VFS: 2},
+            Ethernet.SRIOV_SUBTREE: {
+                Ethernet.SRIOV.DRIVERS_AUTOPROBE: True,
+                Ethernet.SRIOV.TOTAL_VFS: 2,
```
I honestly think that we do not need to patch other tests to add `Ethernet.SRIOV.DRIVERS_AUTOPROBE: True` property if it is not mandatory to specify (at least I can not tell that it is mandatory from the setting that you specified in `rust/src/lib/ifaces/sriov.rs` ).

rjarry replied:
Actually, after setting it to `False` once, it will not reset to `True` implicitly. I haven't looked at the NetworkManager code base but if it is only changing the value when nmstate explicitly asks for it, then `sriov_drivers_autoprobe` will remain set to `0` in sysfs.

I am now wondering if nmstate should force `autoprobe-drivers=1` in NetworkManager if the setting is unset in nmstate.

rjarry replied:
I have removed `Ethernet.SRIOV.DRIVERS_AUTOPROBE: True` as you requested and modified the NetworkManager settings code path to set autoprobe-drivers=true if it is unspecified by nmstate.

liangwen12year replied:
Yes, I think that this way will align better with the behavior in NM.

liangwen12year replied:
```
       sriov.autoprobe-drivers
           Whether to autoprobe virtual functions by a compatible driver.

           If set to NM_TERNARY_TRUE (1), the kernel will try to bind VFs to a compatible driver and if this succeeds a new network interface will be instantiated for each VF.

           If set to NM_TERNARY_FALSE (0), VFs will not be claimed and no network interfaces will be created for them.

           When set to NM_TERNARY_DEFAULT (-1), the global default is used; in case the global default is unspecified it is assumed to be NM_TERNARY_TRUE (1).

           Format: NMTernary (int32)

```

rjarry commented:
@cathay4t I have skipped `verify_sriov()` when `drivers-autoprobe: true` but the integration test still fails... It looks like a timing issue though.

<details>

```
__________________ TestSrIov.test_wait_sriov_vf_been_created ___________________

self = <integration.sriov_test.TestSrIov object at 0x7f3796d947f0>

    def test_wait_sriov_vf_been_created(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {Ethernet.SRIOV.TOTAL_VFS: 2}
                    },
                }
            ]
        }
        try:
            libnmstate.apply(desired_state)
            assertlib.assert_state_match(desired_state)
            vf_ifaces = get_sriov_vf_names(pf_name)
>           assert len(vf_ifaces) == 2
E           AssertionError: assert 1 == 2
E            +  where 1 = len(['ens1f0v1'])

sriov_test.py:220: AssertionError

_______ TestSrIov.test_wait_sriov_vf_been_deleted_when_total_vfs_decrease ______

self = <integration.sriov_test.TestSrIov object at 0x7f3796e308e0>

    def test_wait_sriov_vf_been_deleted_when_total_vfs_decrease(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {Ethernet.SRIOV.TOTAL_VFS: 2}
                    },
                }
            ]
        }
        try:
            libnmstate.apply(desired_state)
            assertlib.assert_state_match(desired_state)
            vf_ifaces = get_sriov_vf_names(pf_name)
>           assert len(vf_ifaces) == 2
E           AssertionError: assert 1 == 2
E            +  where 1 = len(['ens1f0v1'])

sriov_test.py:244: AssertionError

__________________ TestSrIov.test_change_vf_parameters_only ____________________

self = <integration.sriov_test.TestSrIov object at 0x7f3796d6a7c0>, sriov_with_62_vfs = None

    def test_change_vf_parameters_only(self, sriov_with_62_vfs):
        pf_name = _test_nic_name()
        cur_iface = show_only((pf_name,))[Interface.KEY][0]
        iface_infos = [
            {
                Interface.NAME: pf_name,
                Interface.TYPE: InterfaceType.ETHERNET,
                Interface.STATE: InterfaceState.UP,
                Ethtool.CONFIG_SUBTREE: cur_iface[Ethtool.CONFIG_SUBTREE],
                Ethernet.CONFIG_SUBTREE: {
                    Ethernet.SRIOV_SUBTREE: {
                        Ethernet.SRIOV.VFS_SUBTREE: [VF0_CONF, VF1_CONF],
                    }
                },
            },
        ]
        desired_state = {Interface.KEY: iface_infos}
        libnmstate.apply(desired_state)
        vf_ifaces = get_sriov_vf_names(pf_name)
>       assert len(vf_ifaces) == 62
E       AssertionError: assert 2 == 62
E        +  where 2 = len(['ens1f0v0', 'ens1f0v1'])

sriov_test.py:528: AssertionError

___________________ TestSrIov.test_drivers_autoprobe_false _____________________

self = <integration.sriov_test.TestSrIov object at 0x7f3796dbfa60>

    def test_drivers_autoprobe_false(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {
                            Ethernet.SRIOV.DRIVERS_AUTOPROBE: False,
                            Ethernet.SRIOV.TOTAL_VFS: 32,
                        },
                    },
                }
            ]
        }
        libnmstate.apply(desired_state)
        vf_ifaces = get_sriov_vf_names(pf_name)
>       assert len(vf_ifaces) == 0
E       AssertionError: assert 1 == 0
E        +  where 1 = len(['ens1f0v1'])

sriov_test.py:547: AssertionError

_____________ TestSrIov.test_drivers_autoprobe_restore_default _________________

self = <integration.sriov_test.TestSrIov object at 0x7f37965fd610>

    def test_drivers_autoprobe_restore_default(self):
        pf_name = _test_nic_name()
        desired_state = {
            Interface.KEY: [
                {
                    Interface.NAME: pf_name,
                    Ethernet.CONFIG_SUBTREE: {
                        Ethernet.SRIOV_SUBTREE: {Ethernet.SRIOV.TOTAL_VFS: 2},
                    },
                }
            ]
        }
        libnmstate.apply(desired_state)
        vf_ifaces = get_sriov_vf_names(pf_name)
>       assert len(vf_ifaces) == 2
E       AssertionError: assert 1 == 2
E        +  where 1 = len(['ens1f0v1'])

sriov_test.py:563: AssertionError
```

</details>

cathay4t commented on rust/src/lib/nm/settings/sriov.rs:
```diff
@@ -27,6 +27,13 @@ pub(crate) fn gen_nm_sriov_setting(
+        nm_sriov_set.autoprobe_drivers = Some(autoprobe);
+    } else {
+        // if unset, always force to true so that nm restores the kernel default
+        nm_sriov_set.autoprobe_drivers = Some(true);
```
You are overriding current setting even when desired state not mentioned so.

For example, assuming pre-apply has `autoprobe_drivers: false`, when you applying desired state changing total vfs count only, your patch will set `autoprobe_drivers: true` explicitly.

cathay4t replied:
If you want to default `autoprobe_drivers` to be true if not mentioned in desire state and current state, please use the same workflow as https://github.com/nmstate/nmstate/pull/2665

cathay4t commented on rust/src/lib/query_apply/sriov.rs:
```diff
@@ -54,6 +55,17 @@ impl SrIovConfig {
+            .ethernet
+            .as_ref()
+            .and_then(|eth_conf| eth_conf.sr_iov.as_ref())
+            .and_then(|sriov_conf| sriov_conf.drivers_autoprobe.as_ref())
```
Skip the verification if current state has `drivers_autoprobe: false` is wrong, we might pass the verification when desired `drivers_autoprobe: true` but current `drivers_autoprobe: false`.

I think you should skip sriov vf iface verification only when desired state has `drivers_autoprobe: false` or desired not mentioned with current `drivers_autoprobe: false`.

cathay4t commented:
@rjarry Any progress?

rjarry commented:
@cathay4t sorry, I was sidetracked by something else since for the last two weeks. I haven't given up on this PR yet :) Hopefully, I will find some time to make the changes you requested soon.

rjarry commented:
Hi @cathay4t could you have a look at the latest changes? Let me know if what I did is OK. Thanks!

cathay4t commented:
Sure. Will review before June 13 2024.

rjarry commented:
Thanks!

cathay4t commented on rust/src/lib/nispor/ethernet.rs:
```diff
@@ -18,6 +19,29 @@ fn gen_eth_conf(np_iface: &nispor::Iface) -> EthernetConfig {
+                }
+            }
+        }
+    }
```
I am OK to have these code in nmstate project for now. But if possible, please move this code to `nispor` crate at https://github.com/nispor/nispor/

rjarry replied:
I had considered putting the code there but `sriov_drivers_autoprobe` is a PCI level attribute, not a linux network device attribute. It is not exposed by netlink and only via sysfs, it seemed that it didn't have its place in `nispor`.

That makes me think it explains why the integration tests are failing. Since nmstate relies on `nispor` to get the list of VFs associated with a netdev, it will never see VFs created with `drivers_autoprobe = false`. For that, `nmstate` would need to read `sriov_numvfs` in sysfs.

Let me know what you think.

cathay4t replied:
nispor currently holds code for querying `sriov_numvfs` even they are in PCI domain. It is doing all heavy lifting on querying network kernel states, nmstate is a consumer here and trust nispor provide accurate data regardless its source.

In nispor, once found `drivers_autoprobe = false`, it could skip searching for VF NIC names. So ideally, nispor should hold the sysfs access code. It could be done later.

Still begging for servers. :(

cathay4t commented on rust/src/lib/query_apply/sriov.rs:
```diff
@@ -54,6 +57,23 @@ impl SrIovConfig {
                 }
             };
 
+        if let Some(desired_autoprobe) = self.drivers_autoprobe {
```
```rust
// Skip VF NIC searching as `drivers_autoprobe` disabled.
if self.drivers_autoprobe  == Some(false) {
    return Ok(());
}
```

rjarry replied:
Ack.

cathay4t reviewed (changes requested):
The code looks OK to me, only some lints.

Please give me one more day to find a real SRIOV server to test it.

cathay4t commented:
The `test_drivers_autoprobe_false` does not clean up which cause follow up test case `test_drivers_autoprobe_restore_default` fails.

After applying this patch, your integration test case works well:

```patch
diff --git a/tests/integration/sriov_test.py b/tests/integration/sriov_test.py
index 4b40064a..0907dbd8 100644
--- a/tests/integration/sriov_test.py
+++ b/tests/integration/sriov_test.py
@@ -527,7 +527,7 @@ class TestSrIov:
         vf_ifaces = get_sriov_vf_names(pf_name)
         assert len(vf_ifaces) == 62
 
-    def test_drivers_autoprobe_false(self):
+    def test_drivers_autoprobe_false(self, disable_sriov):
         pf_name = _test_nic_name()
         desired_state = {
             Interface.KEY: [
@@ -546,7 +546,7 @@ class TestSrIov:
         vf_ifaces = get_sriov_vf_names(pf_name)
         assert len(vf_ifaces) == 0
 
-    def test_drivers_autoprobe_restore_default(self):
+    def test_drivers_autoprobe_restore_default(self, disable_sriov):
         pf_name = _test_nic_name()
         desired_state = {
             Interface.KEY: [
```

Let me amend your patch and rebase it.

cathay4t commented on examples/eth1_with_sriov.yml:
```diff
@@ -5,6 +5,7 @@ interfaces:
     state: up
     ethernet:
       sr-iov:
+        drivers-autoprobe: false
```
I guess most of SRIOV user are expecting a VF been created. Hence I would suggest remove this change to example YAML.

cathay4t reviewed (approved):
Tested on ixgbe SRIOV NIC card.

cathay4t commented:
Since you are using main branch for PR, I cannot amend your PR. I will create new PR after merged.

cathay4t committed:
Merge branch 'base' into base

rjarry commented:
Thanks for merging! I can create a new PR to fix the remaining issues. Do you prefer to do it yourself?

cathay4t commented:
Already done and shipped in 2.2.33.

rjarry commented:
Awesome, thanks. By the way, I have added support for sysfs sriov attributes in nispor as you requested: https://github.com/nispor/nispor/pull/274
